denops/glance/main.ts

```diff
@@ -20,37 +35,63 @@ export async function main(denops: Denops) {
   }
 
   async function update() {
+    renderer = await ensureRenderer();
+    server = await ensureServer();
     const lines = await fn.getline(denops, 1, "$");
     const content = lines.join("\n");
     const document = await renderer.render(content);
     const pos = await fn.getpos(denops, ".");
     server.send("update", { document, line: pos[1] });
   }
-  const port = (await g.get(denops, "glance#server_port", 8765))!;
-  const plugins = (await g.get(denops, "glance#markdown_plugins", []))!;
-  const html = (await g.get(denops, "glance#markdown_html", false))!;
-  const breaks = (await g.get(denops, "glance#markdown_breaks", false))!;
-  const linkify = (await g.get(denops, "glance#markdown_linkify", false))!;
-  const defaultStylesheet = `#root {margin: 50px auto; width: min(700px, 90%);}`;
-  const stylesheet = (await g.get(denops, "glance#stylesheet", defaultStylesheet))!;
-  const defaultConfigPath = new URL("./config.ts", import.meta.url).toString();
-  const configPath = (await g.get(denops, "glance#config", defaultConfigPath))!;
-  const { createMarkdownRenderer } = await import(configPath);
-  const renderer = new MarkdownRenderer();
-  await renderer.initialize({ html, breaks, linkify, plugins, createMarkdownRenderer });
-  const server = new Server({ onOpen: update, readFile, stylesheet });
+
+  async function ensureOptions() {
+    if (options) return options;
+    const defaultStylesheet = `#root {margin: 50px auto; width: min(700px, 90%);}`;
+    const defaultConfigPath = new URL("./config.ts", import.meta.url).toString();
+    options = {
+      port: await g.get(denops, "glance#server_port", 8765),
+      plugins: await g.get(denops, "glance#markdown_plugins", []),
+      html: await g.get(denops, "glance#markdown_html", false),
+      breaks: await g.get(denops, "glance#markdown_breaks", false),
+      linkify: await g.get(denops, "glance#markdown_linkify", false),
+      defaultStylesheet,
+      stylesheet: await g.get(denops, "glance#stylesheet", defaultStylesheet),
+      defaultConfigPath,
+      configPath: await g.get(denops, "glance#config", defaultConfigPath),
+    };
+    return options;
+  }
+
+  async function ensureRenderer() {
+    if (renderer) return renderer;
+    options = await ensureOptions();
+    const { createMarkdownRenderer } = await import(options.configPath);
+    renderer = new MarkdownRenderer();
+    const { html, breaks, linkify, plugins } = options;
+    await renderer.initialize({ html, breaks, linkify, plugins, createMarkdownRenderer });
+    return renderer;
+  }
+
+  async function ensureServer() {
+    if (server) return server;
+    options = await ensureOptions();
+    server = new Server({ onOpen: update, readFile, stylesheet: options.stylesheet });
```

Comment:
This is not a request for revision. Don't we need to use `options?.` to avoid type-error because the type of `options`  is `Options | undefined` ?

Reply:
ensureOptions() always returns `Options` so there should be no problem :)
My nvim-lsp infers `options` in L78 as `Options`.

But sure, the implementations of ensureXXX() here are not great.
I wanted to define user-defined type guards but async-await can not be used.

Reply:
Awesome. Thank you for answering my question.